templates/generic_show_form.html.twig

```diff
@@ -53,635 +53,528 @@
        'rand': rand,
    } %}
 
+    {# NOTE: Post code field named differently for Suppliers. Placed after town to maintain field order from 9.5)#}
+    {% set default_field_order = [
+        'name', 'firstname', 'template_name', '_template_is_active', 'states_id', item.getForeignKeyField(), 'is_helpdesk_visible',
+        '_dc_breadcrumbs', 'locations_id', 'item_type', 'itemtype', 'date_domaincreation', item.getTypeForeignKeyField(),
+        'usertitles_id', 'registration_number', 'phone', 'phone2', 'phonenumber', 'mobile', 'fax', 'website', 'email',
+        'address', 'postalcode', 'town', 'postcode', 'state', 'country', 'date_expiration', 'ref', 'users_id_tech',
+        'manufacturers_id', 'groups_id_tech', item.getModelForeignKeyField(), 'contact_num', 'serial', 'contact', 'otherserial',
+        'sysdescr', 'snmpcredentials_id', 'users_id', 'is_global', 'size', 'networks_id', 'groups_id', 'uuid', 'version',
+        'comment', 'ram', 'alarm_threshold', 'brand', 'begin_date', 'autoupdatesystems_id', 'pictures', 'is_active', 'last_boot'
+    ] %}
+    {# Support custom fields for generic assets #}
+    {% set custom_fields = custom_fields|default({}) %}
+    {# Set real field older using 'field_order' provided and any fields missing added to the end of the order array #}
+    {% set field_order = field_order|default(default_field_order) %}
+    {% set field_order = field_order|merge(array_diff(default_field_order, field_order)) %}
+    {# Add missing custom fields to end of the order array #}
+    {% set custom_fields_keys = custom_fields|keys %}
+    {% set field_order = field_order|merge(array_diff(custom_fields_keys, field_order)) %}
+    {# Remove fields that don't apply #}
+    {% set field_order = field_order|filter((f) => f starts with '_' or item.isField(f)) %}
+
+    {# Exclude specified fields #}
+    {% set fields_excluded = fields_excluded|default([]) %}
+    {% set field_order = field_order|filter((f) => f not in fields_excluded) %}
+
```

Comment:
That seems like a lot of computation inside a view.
Maybe it could be computed by a testable method on PHP's side ?

Reply:
There isn't really a lot of logic here and it makes sense to handle it all at this level. At best, this would be replaced by a `call` to a PHP method, but that doesn't seem strictly needed.

Reply:
This code is pretty hard to read and it cannot be tested. You could probably move it inside a `CommonDBTM::getFormFields()` that could be overriden by the `Glpi\Asset\Asset` class.

Reply:
I didn't want to define any kind of `getFormFields` method in the PHP class until we decide to properly define fields, but now it is just an array of names. Will need to figure out a proper schema later and handling changing that.